### PR TITLE
Setting workspace_owner_object_id when creating workspaces

### LINKED ISSUE #220: AD app registration created when new workspace is created
## Description

As a TRE administrator
I want to create workspaces without providing identity provider parameters such as application id
So that I don't have to bother our cloud admin asking her to create the application registration for me first.

## Acceptance criteria

- [x] #1706
- [x] #211 
- [x] #1797
- [x] #1928
- [x] App registration deleted when the workspace is deleted (Terraform is a good option for creation/deletion)
- [ ] Documentation updated with environment variable and permissions for automatic app registration
- [x] #831
- [x] #1716

## Design 

Workspace templates must only change resources it itself creates. Changes done to management/core services such as Azure Firewall, Shared Service such as source mirror, Azure Active Directory etc. must be delegated.

Support for alternative identity providers other than Azure Active Directory for application registration must also be supported. It should be configurable which identity provider is used at TRE creation time (not part of this Story). Meaning the same identity provider will be used in all Workspaces within the same TRE instance.

Therefore, the responsible for creating application registrations is implemented into the TRE management API.

### Permissions

To be able to automatically create app registration the [Application.ReadWrite.All](https://docs.microsoft.com/en-us/graph/permissions-reference#delegated-permissions-4) permission must be granted to the TRE management API Service Principal

Azure AD is managed via the Microsoft Graph API.

### Changes to setup 

A new environment variable `auto_app_registration = [TRUE|FALSE] (default=FALSE)` indication if automatic app registration is enabled when setting up a TRE instance.
If the environment variable is set to true, then the Application.ReadWrite.All permission must be granted.

### Changes to the TRE API

Currently the following parameters are required in the REST API to create a Workspace

```JSON
{
  "workspaceType": "tre-workspace-vanilla",
  "properties": {
    "display_name": "the workspace display name",
    "description": "workspace description",
    "client_id": "9d52b04f-89cf-47b4-868a-e12be7133b36"
  }
}
```

The `app_id` should be made optional, but only if the `auto_app_registration` variable is set to TRUE.

The abstraction over identity provider is implemented in the [AccessService](https://github.com/microsoft/AzureTRE/blob/develop/management_api_app/services/access_service.py) class. This class must be extended to include an application registration method.

Incorporate and reuse as much as possible from the [manual app registration script](https://github.com/microsoft/AzureTRE/blob/develop/scripts/workspace-app-reg.py).

Use the AAD API to create the app registration ([Create application - Microsoft Graph v1.0 | Microsoft Docs](https://docs.microsoft.com/en-us/graph/api/application-post-applications?view=graph-rest-1.0&tabs=http))

### PR DESCRIPTION
**Pull request by ross-p-smith**
## What is being addressed

1. We need to know the workspace owner when we are creating a workspace, so that we can grant it the WorkspaceOwner permission so that the tests work correctly.

NB - This is only for workspaces that have a value of "auto_create" for the client_id of the workspace. Fixes #220 

2. Guacamole uses the `api://<identifier_Uri>` to authenticate with the Workspace AAD Application. This can no longer be derived by client_id, but can be configurable; especially if you auto_create the Workspace AD App. Therefore this value is passed in to the swagger when we create the Guacamole workspace service, so it can be used as the scope/audience.

## Outstanding
- At the moment, you cannot pass in workspace_owner_object_id in the Swagger. This will be another PR

## How is this addressed

- At the moment, the current logged in user of the API is passed into the workspace creation process and that user will be the WorkspaceOwner
- Updated the Terraform to assign `WorkspaceOwner` role permission to the passed in object id.
- When we create the workspace we need to wait until the key vault DNS is ready before we add secrets to it. Wait on the null resource provider to complete
- Guacamole service now requires the workspace_identifier_uri to be passed in on creation. This can be found in the workspace properties under scope_id.

### CODE REVIEW

File: templates/workspaces/base/terraform/workspace.tf [hunk 1]
```diff
@@ -31,6 +31,6 @@ module "aad" {
   depends_on = [
     azurerm_key_vault_access_policy.deployer,
     azurerm_key_vault_access_policy.resource_processor,
-    azurerm_private_endpoint.kvpe
+    null_resource.wait_for_dns_vault
```
Comment (tamirkamara) [inline templates/workspaces/base/terraform/workspace.tf:34]: What is this for?

File: e2e_tests/test_workspace_services.py [hunk 2]
```diff
@@ -56,7 +56,7 @@ async def test_create_guacamole_service_into_base_workspace(admin_token, verify)
 
 @pytest.mark.extended_aad
 @pytest.mark.timeout(3000)
-async def test_create_guacamole_service_into_aad_workspace(admin_token, workspace_owner_token, verify) -> None:
+async def test_create_guacamole_service_into_aad_workspace(admin_token, verify) -> None:
```
Comment (tamirkamara) [inline e2e_tests/test_workspace_services.py:59]: Why have you changed the token to the admin one? A workspace service needs to be deployed by the owner... 
Or, am I missing something?

File: templates/workspaces/base/terraform/workspace.tf (hunk 1 above)
Comment (ross-p-smith) [reply templates/workspaces/base/terraform/workspace.tf:34]: When you write items to the keyvault, you have to do it over the private endpoint. However, there is a delay between the private endpoint being created and then that ip address being added to the DNS. This is just a nasty sleep for 2 minutes for the DNS to resolve.

File: e2e_tests/test_workspace_services.py (hunk 2 above)
Comment (ross-p-smith) [reply e2e_tests/test_workspace_services.py:59]: The E2E tests create the workspace dynamically, and to do that they need the admin token as that route is an admin only route. We used to have a fixture that would get the workspace token by using a convention `api://<app_reg_id`; that convention is not true anymore as we need to read a property of the workspace (scope_id) to tell us what that value is. Once we have that value, then we dynamically get the workspace_token scoped to that application and then use the workspace_owner token from then on in.

Basically, what used to be a test fixture, is now a method called **get_workspace_owner_token** that you call from your test.

File: api_app/tests_ma/test_db/test_repositories/test_workpaces_repository.py [hunk 3]
```diff
@@ -201,3 +203,17 @@ def test_automatically_create_application_registration_returns_false(workspace_r
...
+    expected_object_id = "Expected"
+
+    assert workspace_repo.get_workspace_owner(dictToTest, expected_object_id) is expected_object_id
+
+
+def test_workspace_owner_is_not_overwritten_if_not_present_in_workspace_properties(workspace_repo):
+    dictToTest = {"workspace_owner_object_id": "Expected"}
+    not_expected_object_id = "Not Expected"
+
+    assert workspace_repo.get_workspace_owner(dictToTest, not_expected_object_id) == "Expected"
```
Comment (LizaShak) [inline api_app/tests_ma/test_db/test_repositories/test_workpaces_repository.py:219]: I think it will be also good to have a positive test where you do have a workspace owner specified in the properties to make sure it is set correctly

File: e2e_tests/helpers.py [hunk 4]
```diff
@@ -246,4 +246,4 @@ async def get_workspace_owner_token(admin_token, workspace_id, verify) -> Option
 
         token = responseJson["access_token"]
 
-        return token if (response.status_code == status.HTTP_200_OK) else None
+        return token, scope_uri
```
Comment (LizaShak) [inline e2e_tests/helpers.py:249]: What should happen now if we get a response which is not 200?

File: templates/workspace_services/guacamole/version.txt [hunk 5]
```diff
@@ -1 +1 @@
-__version__ = "0.3.0"
+__version__ = "0.3.1"
```
Comment (LizaShak) [inline templates/workspace_services/guacamole/version.txt:1]: Shouldn't it be the same as in : templates/workspace_services/guacamole/porter.yaml? Mmmm I wonder how did it work until now, or maybe I  am missing something....

File: api_app/tests_ma/test_db/test_repositories/test_workpaces_repository.py (hunk 3 above)
Comment (ross-p-smith) [reply api_app/tests_ma/test_db/test_repositories/test_workpaces_repository.py:219]: Maybe I misunderstood what you mean, but there are 2 new tests
1) Does not have `workspace_owner_object_id` in a dictionary (workspace properties), so therefore returns the default.
2) Does have `workspace_owner_object_id` in a dictionary (workspace properties), so ignores the default

File: e2e_tests/helpers.py (hunk 4 above)
Comment (ross-p-smith) [reply e2e_tests/helpers.py:249]: An exception is thrown a few lines above if there is no access token in the response. It would have never met a false condition `response.status_code == status.HTTP_200_OK`.

**Review by LizaShak — APPROVED** [review]
